packages/shell-api/src/stream-processor.ts

```diff
@@ -0,0 +1,121 @@
+import { promisify } from 'util';
+
+import type { Document } from '@mongosh/service-provider-core';
+
+import type Mongo from './mongo';
+import { asPrintable } from './enums';
+import {
+  ShellApiWithMongoClass,
+  returnsPromise,
+  shellApiClassDefault,
+} from './decorators';
+
+import type { Streams } from './streams';
+
+const sleep = promisify(setTimeout);
+
+@shellApiClassDefault
+export default class StreamProcessor extends ShellApiWithMongoClass {
+  constructor(public _streams: Streams, public name: string) {
+    super();
+  }
+
+  get _mongo(): Mongo {
+    return this._streams._mongo;
+  }
+
+  [asPrintable]() {
+    return `Atlas Stream Processor: ${this.name}`;
+  }
+
+  @returnsPromise
+  async start() {
+    const result = await this._streams._runCommand({
+      startStreamProcessor: 1,
+      name: this.name,
+    });
+    return result;
+  }
+
+  @returnsPromise
+  async stop() {
+    const result = await this._streams._runCommand({
+      stopStreamProcessor: 1,
+      name: this.name,
+    });
+    return result;
+  }
+
+  @returnsPromise
+  async drop() {
+    const result = await this._streams._runCommand({
+      dropStreamProcessor: 1,
+      name: this.name,
+    });
+    return result;
+  }
+
+  @returnsPromise
+  async stats(options: Document = {}) {
+    return this._streams._runCommand({
+      getStreamProcessorStats: 1,
+      name: this.name,
+      ...options,
+    });
+  }
+
+  @returnsPromise
+  async sample(options: Document = {}) {
+    const r = await this._streams._runCommand({
+      startSampleStreamProcessor: 1,
+      name: this.name,
+      ...options,
+    });
+
+    if (r.ok !== 1) {
+      return r;
+    }
+
+    return this._sampleFrom(r.cursorId as number);
+  }
+
+  @returnsPromise
+  async _sampleFrom(cursorId: number) {
+    let currentCursorId = cursorId;
+    // keep pulling until end of stream
+    while (currentCursorId !== 0) {
+      const res = await this._streams._runCommand({
+        getMoreSampleStreamProcessor: 1,
+        name: this.name,
+        cursorID: currentCursorId,
+      });
+
+      if (res.ok !== 1) {
+        return res;
+      }
+
+      currentCursorId = res.cursorId;
+
+      // print fetched documents
+      for (const doc of res.messages) {
+        // eslint-disable-next-line no-console
+        console.log(doc);
```

Comment:
Unless this is a WIP, you may want to use something like `this._instanceState.shellApi.print(...)` instead, which is wired with the presentation layer of the REPL.

As far as i know `console.log` wouldn't work when the shell is embedded in Compass, VSCode, Jetbrains etc .. and could probably have unwanted side effects for the CLI too, like not applying user preferences for the output .. although i'm not sure about that.

Reply:
Good call. Didn't know about `this._instanceState.shellApi.print(...)`, will update to use that.

Reply:
Have changed to use `printjson`.
```ts
      for (const doc of res.messages) {
        await this._instanceState.shellApi.printjson(doc);
      }
```